### ISSUE TITLE
Bug: [no-invalid-void-type] allowInGenericTypeArguments does not support array or union

### ISSUE THREAD

**Issue by HolgerJeromin** [typescript-eslint/typescript-eslint#4998]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=CYUwxgNghgTiAEEQBd4AcYHsC2AZAlgM7IBc8A2gApbZEgA8AbpvsAHwC6A3ALABQSVBhzwS1HHSYtg5DgB9mrNvF598AMwAUAWSjIAFgDoYUAHbAcmgJRWA3v3iP0NeAF5442oRCGoECJrCeETIVqoAviAQ3vZ8Ts4i7kEExOQADNz84fxAA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZaEBuAhvgCZZVlqPTwgBcA2iHCijIoQAGlB1cZAO4BBBABUAFmgQAFGihoBbOBH5cIKAK5wRICdICSBAOJwC-NEnmw4MlAHMjOoggPG4AF8AXUCgA&tsConfig=N4XyA

### Repro Code

```typescript
declare let promList: [Promise<void>];
let prom :Promise<void[]|void> ;
if(Math.random()){
    prom = Promise.all(promList);
}else{
    prom = promList[0];
}
// related
declare let oneProm: Promise<void>;
const prom2: Promise<[void]>= Promise.all([oneProm]);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
  '@typescript-eslint/no-invalid-void-type': ['error', { allowInGenericTypeArguments: true }],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expected that I can use void in the type as part of an array or union with an array.
This very type is inferred within typescript itself when no type is provided.

### Actual Result

> void is only valid as a return type or generic type variable or the type of a `this` parameter.ESLint(@typescript-eslint/no-invalid-void-type)

For all types provided.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.24.0` |
| `@typescript-eslint/parser`            | `5.24.0` |
| `TypeScript`                           | `4.6.2` |
| `ESLint`                               | `8.15.0` |

**Comment by Solo-steven** [comment]
Hi @JoshuaKGoldberg, I am very interested in solving this rule, but I want to discuss with you how to handle it to insure I don't misunderstand this issue.
I think the way to solve this issue is to accept void as part of a tuple or an array in generic type when `allowInGenericTypeArguments` is true. like below :
<img width="735" alt="Screenshot 2023-01-14 at 1 23 24 PM" src="https://user-images.githubusercontent.com/67428009/212457327-2f19db83-e8cb-4e36-9174-1825d13e58fb.png">

**Comment by Josh-Cena** [comment]
I'm going to close this as a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/8113. Let's discuss there instead!
